core/offchain/impl/offchain_worker_impl.cpp

```diff
@@ -0,0 +1,340 @@
+    if (result.has_error()) {
+      SL_WARN(log_,
+              "Can't compare-and-set value in storage: {}",
+              result.error().message());
+    }
+    return result.value();
+  }
+
+  outcome::result<common::Buffer> OffchainWorkerImpl::localStorageGet(
+      StorageType storage_type, const common::Buffer &key) {
+    auto &storage = getStorage(storage_type);
+    auto result = storage.get(key);
+    if (result.has_error()) {
+      SL_WARN(log_, "Can't get value in storage: {}", result.error().message());
+      return result.as_failure();
+    }
+    return std::move(result.value());
+  }
+
+  Result<RequestId, Failure> OffchainWorkerImpl::httpRequestStart(
+      HttpMethod method, std::string_view uri, common::Buffer meta) {
+    auto request_id = ++request_id_;
+
+    auto request = std::make_shared<HttpRequest>(request_id);
+
+    if (not request->init(method, uri, meta)) {
+      return Failure();
+    }
+
+    auto is_emplaced = requests_.emplace(request_id, std::move(request)).second;
+
+    if (is_emplaced) {
+      return request_id;
+    } else {
+      return Failure();
+    }
+  }
+
+  Result<Success, Failure> OffchainWorkerImpl::httpRequestAddHeader(
+      RequestId id, std::string_view name, std::string_view value) {
+    auto it = requests_.find(id);
+    if (it == requests_.end()) {
+      return Failure();
+    }
+    auto &request = it->second;
+
+    request->addRequestHeader(name, value);
+
+    return Success();
+  }
+
+  Result<Success, HttpError> OffchainWorkerImpl::httpRequestWriteBody(
+      RequestId id, common::Buffer chunk, std::optional<Timestamp> deadline) {
+    std::optional<std::chrono::milliseconds> timeout = std::nullopt;
+    if (deadline.has_value()) {
+      timeout = std::chrono::duration_cast<std::chrono::milliseconds>(
+          std::chrono::milliseconds(deadline.value())
+          - clock_->now().time_since_epoch());
+    }
+
+    auto it = requests_.find(id);
+    if (it == requests_.end()) {
+      return HttpError::InvalidId;
+    }
+    auto &request = it->second;
+
+    auto result = request->writeRequestBody(chunk, timeout);
+
+    return result;
+  }
+
+  std::vector<HttpStatus> OffchainWorkerImpl::httpResponseWait(
+      const std::vector<RequestId> &ids, std::optional<Timestamp> deadline) {
+    std::vector<HttpStatus> result;
+    result.reserve(ids.size());
+
+    for (auto id : ids) {
+      auto it = requests_.find(id);
+      if (it == requests_.end()) {
+        result.push_back(0);  // Invalid identifier
+      }
+      auto &request = it->second;
+
+      HttpStatus status;
+      while ((status = request->status()) == 0) {
+        if (deadline.has_value()
+            and (clock_->zero() + std::chrono::milliseconds(deadline.value()))
+                    < clock_->now()) {
+          break;
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
+      }
+
+      result.push_back(status ? status : HttpStatus(10));
```

Comment:
I think it's better to make HttpStatus(10) a named constant.

Reply:
Good idea! Implemented.